### PR TITLE
Fix automatic addition of protocol in discovery endpoint 

### LINKED ISSUE #11609: [BUG] Automatic addition of protocol broken in #11512
**Describe the bug**

Coming from https://github.com/opensearch-project/OpenSearch/pull/11512, the check examines the host for `http` which is broken for `httpserver.example.com`. See https://github.com/opensearch-project/OpenSearch/pull/11512#issuecomment-1854313297.

**Expected behavior**

Check for full protocol, `http://` and `https://`.

### PR DESCRIPTION
**Pull request by heyanlong**
### Description
Fix #11609 

### Related Issues
Resolves #11609 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [ ] ~Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))~
- [x] Commits are signed per the DCO using --signoff
- [ ] ~Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by kkmr — APPROVED** [review]

File: plugins/discovery-ec2/src/test/java/org/opensearch/discovery/ec2/AwsEc2ServiceImplTests.java [hunk 1]
```diff
@@ -222,6 +222,8 @@ public void testGetFullEndpointWithoutScheme() {
         String endpoint = awsEc2ServiceImpl.getFullEndpoint(clientSettings.endpoint);
         assertEquals("https://ec2.us-west-2.amazonaws.com", endpoint);
 
+        assertEquals("https://httpserver.example.com", awsEc2ServiceImpl.getFullEndpoint("httpserver.example.com"));
```
Comment (dblock) [inline plugins/discovery-ec2/src/test/java/org/opensearch/discovery/ec2/AwsEc2ServiceImplTests.java:225]: Add a test for `http://` as well to exercise that code path.

**Review by dblock — COMMENTED** [review]
Before we merge this, any reason to hand-roll our own implementation vs. using something like `HttpHost.create(endpoint).getHostName()`?

**Comment by heyanlong** [comment]
> Before we merge this, any reason to hand-roll our own implementation vs. using something like `HttpHost.create(endpoint).getHostName()`?

Yes, we can use `HttpHost.create(endpoint).getScheme()` to check if it has scheme, but I don't think this simple validation requires using a third party implementation

**Comment by dblock** [comment]
> > Before we merge this, any reason to hand-roll our own implementation vs. using something like `HttpHost.create(endpoint).getHostName()`?
> 
> Yes, we can use `HttpHost.create(endpoint).getScheme()` to check if it has scheme, but I don't think this simple validation requires using a third party implementation

That's fine here. Generally I prefer well tested libraries that implement an RFC because, I think, you'll agree that this exact PR is a demonstration of the fact that validation may not be totally trivial. This fixes a bug introduced by rolling out custom validation.

**Review by dblock — APPROVED** [review]

**Comment by heyanlong** [comment]
> > > Before we merge this, any reason to hand-roll our own implementation vs. using something like `HttpHost.create(endpoint).getHostName()`?
> > 
> > 
> > Yes, we can use `HttpHost.create(endpoint).getScheme()` to check if it has scheme, but I don't think this simple validation requires using a third party implementation
> 
> That's fine here. Generally I prefer well tested libraries that implement an RFC because, I think, you'll agree that this exact PR is a demonstration of the fact that validation may not be totally trivial. This fixes a bug introduced by rolling out custom validation.

It is like this, for stricter validation, libraries are a better choice.

**Comment by dblock** [comment]
> It is like this, for stricter validation, libraries are a better choice.

@heyanlong feel free to PR a change that uses the library! we can compare looking at the code

**Pull request merged**